anchor-scope name match is tree-scoped

fantasai opened the pull request:
#### 4774b6e10334f22c7d810df5f7c0eadca642356d
<pre>
anchor-scope name match is tree-scoped
<a href="https://bugs.webkit.org/show_bug.cgi?id=293766">https://bugs.webkit.org/show_bug.cgi?id=293766</a>
<a href="https://rdar.apple.com/152271352">rdar://152271352</a>

Reviewed by Tim Nguyen.

1. Use style-scoped names for anchor-name by storing scope info with anchor-name.
2. Perform scoped name matches when checking anchor-scope in AnchorPositionEvaluator.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-shadow-all-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-shadow-names-expected.txt:
Update tests to passing!

* Source/WebCore/rendering/style/NameScope.h:
(WebCore::operator==):
(WebCore::operator&lt;&lt;):
Introduce a scopeOrdinal data member to NameScope, to represent scoped names.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::anchorScopeForAnchorName):
(WebCore::Style::isAcceptableAnchorElement):
(WebCore::Style::findLastAcceptableAnchorWithName):
Update anchor-scope name matching to account for style scope.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertNameScope):
Update NameScope builder to capture the style scope.

* Source/WebCore/style/StyleScopeOrdinal.h:
(WebCore::Style::operator&lt;&lt;):

Canonical link: <a href="https://commits.webkit.org/297285@main">https://commits.webkit.org/297285@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b5c8a570311ec37f5c74dd9258ce5f7799db161

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111165 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117196 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61433 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113127 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39413 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84506 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114112 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25173 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100090 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64952 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24519 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/18236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61016 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/94554 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18299 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120197 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38214 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28398 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93442 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38590 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96366 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93267 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23766 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38355 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16109 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34199 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38103 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43580 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37768 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41101 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39470 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/style/NameScope.h:
```diff
@@ -40,23 +40,25 @@ struct NameScope {
-    switch (lhs.type) {
-    case NameScope::Type::None:
-        return rhs.type == NameScope::Type::None;
+struct ScopedNameScope : public NameScope {
```
I don't think separating the two adds anything but complexity, non-scoped is usually going to be an exception usually.

tuankiet65 replied:
I guess the reason for the separation is `operator==` difference.

If two scope ordinals are equal to each other, does it mean they belong to the same scope? I vaguely remember it wasn't, you have to resolve it first (using `Scope::forOrdinal`)

Anyway I'd also rather we have one thing and not two.

fantasai replied:
I think the equality operator is mainly used for style diffing and invalidation, so I think it won't make a difference practically speaking.

Let's go with the one thing version. :)

nt1m reviewed (approved):
r=me but please merge back ScopedNameScope into NameScope

tuankiet65 reviewed (commented):
Looks good.

I'm writing down my understanding of this patch to make sure I understand it fully, correct me if I'm wrong somewhere:
* an anchor name is tree-scoped, as represented by `ResolvedNameScope`: an anchor name has a string name and the scope it belongs to
* anchor-scope is tree-scoped too, meaning the name it scopes must also come from the same tree scope as the anchor-scope
* The old way of doing name matching is wrong because we only look at the string name and don't check that the anchor name comes from the same tree scope as anchor-scope
* This patch fixes that

fantasai committed:
anchor-scope name match is tree-scoped
https://bugs.webkit.org/show_bug.cgi?id=293766
rdar://152271352

Reviewed by Tim Nguyen.

1. Use style-scoped names for anchor-name by storing scope info with anchor-name.
2. Perform scoped name matches when checking anchor-scope in AnchorPositionEvaluator.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-shadow-all-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scope-shadow-names-expected.txt:
Update tests to passing!

* Source/WebCore/rendering/style/NameScope.h:
(WebCore::operator==):
(WebCore::operator<<):
Introduce a scopeOrdinal data member to NameScope, to represent scoped names.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::anchorScopeForAnchorName):
(WebCore::Style::isAcceptableAnchorElement):
(WebCore::Style::findLastAcceptableAnchorWithName):
Update anchor-scope name matching to account for style scope.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertNameScope):
Update NameScope builder to capture the style scope.

* Source/WebCore/style/StyleScopeOrdinal.h:
(WebCore::Style::operator<<):

Canonical link: https://commits.webkit.org/297285@main
